### PR TITLE
Attach event listeners upon attribute changes

### PR DESCRIPTION
**Pull request by ZombieRaccoon**
I implemented the feature we discussed in #188, along with a couple of tiny improvements that presented themselves along the way.

A couple of notes about the end result:
* I went with a simple linear search in EventDispatcher::DetachEvent, since I believe that only a relatively high listener count would warrant using binary search and I don't think that a UI library can reach that high a listener count for it to make the extra effort worthwhile.
* In Element::OnAttributeChange it's necessary to traverse the whole dictionary to find attributes that start with "on", so for effectiveness' sake I converted the entire thing to build off of that.

### CODE REVIEW

File: Source/Core/EventDispatcher.h [hunk 1]
```diff
@@ -39,10 +39,12 @@ class EventListener;
 struct CollectedListener;
 
 struct EventListenerEntry {
-	EventListenerEntry(EventId id, EventListener* listener, bool in_capture_phase) : id(id), in_capture_phase(in_capture_phase), listener(listener) {}
+	EventListenerEntry(const EventId id, EventListener* listener, const bool in_capture_phase, const ListenerOrigin listener_origin) : id(id), in_capture_phase(in_capture_phase), listener(listener), listener_origin(listener_origin) {}
+
 	EventId id;
 	bool in_capture_phase;
 	EventListener* listener;
+	ListenerOrigin listener_origin;
```
Comment (mikke89) [inline Source/Core/EventDispatcher.h:47]: I think we might as well move this over the listener member, that way we save a pointer size per entry.

File: Include/RmlUi/Core/Event.h [hunk 2]
```diff
@@ -43,6 +43,7 @@ struct EventSpecification;
 
 enum class EventPhase { None, Capture = 1, Target = 2, Bubble = 4 };
 enum class DefaultActionPhase { None, Target = (int)EventPhase::Target, TargetAndBubble = ((int)Target | (int)EventPhase::Bubble) };
+enum class ListenerOrigin { Attribute, Unknown };
```
Comment (mikke89) [inline Include/RmlUi/Core/Event.h:46]: If we make this private, we can move it to EventDispatcher.h.

File: Source/Core/EventDispatcher.cpp [hunk 3]
```diff
@@ -61,36 +61,40 @@ EventDispatcher::~EventDispatcher()
...
+		{
+			const auto existing_listener_it = std::find_if(range.first, range.second, [](const auto& entry)	{ return entry.listener_origin == ListenerOrigin::Attribute; });
+			if(existing_listener_it != range.second)
+			{
+				DetachEvent(existing_listener_it->id, existing_listener_it->listener, existing_listener_it->in_capture_phase);
+				insertion_position = std::upper_bound(listeners.cbegin(), listeners.cend(), entry, CompareIdPhase());
+			}
+		}
+
+		listeners.emplace(insertion_position, entry);
```
Comment (mikke89) [inline Source/Core/EventDispatcher.cpp:84]: Don't think this is working entirely as intended? If an entry already exists, the first `if` will be skipped, and so the listener is not replaced as I think is the intention here. Also, the nested `find_if` will necessarily return the same as `matching_entry_it`, so I don't see the point there.

**Review by mikke89 — COMMENTED** [review]
Thank you! Should be a nice little improvement. 

I'm not sure it's working entirely correct in terms of replacing old `on...` attributes, see my comments for details. And I think we should also remove the listener when users remove the `on...` attribute.

We should add some tests to make sure it works as intended to the `UnitTests` suite, so that we know replacement and removal works properly. I'm also interested to see if the `onload` attribute is fired properly.

I did benchmark this just to be sure and I couldn't measure any regression, so that is all good!

**Comment by ZombieRaccoon** [comment]
Trying to implement removal and taking the ability to set listener origin off the public interface actually gave me an idea for a different approach, which I think you'll also like. I'll push the commit soon.

File: Source/Core/EventDispatcher.cpp (hunk 3 above)
Comment (ZombieRaccoon) [reply Source/Core/EventDispatcher.cpp:84]: I tested replacement using ```SetAttribute``` locally and it worked. The reason it did is that each time a new listener is created, which means that even though they would fall into the same equal range according to their ```eventId``` and ```phase```, but the event listeners would ultimately make them different. Hence the need for the ```std::find_if```, since within the same equal range we must look for the one that actually originates from an attribute.

File: Include/RmlUi/Core/Element.h [hunk 4]
```diff
@@ -470,14 +470,14 @@ class RMLUICORE_API Element : public ScriptInterface, public EnableObserverPtr<E
...
-	/// @lifetime The added listener must stay alive until after the dispatched call from EventListener::OnDetach(). This occurs
-	///     eg. when the element is destroyed or when RemoveEventListener() is called with the same parameters passed here.
+	/// @warning The added listener must stay alive until after the dispatched call from EventListener::OnDetach().
+	/// This occurs eg. when the element is destroyed or when RemoveEventListener() is called with the same parameters passed here.
 	void AddEventListener(const String& event, EventListener* listener, bool in_capture_phase = false);
 	/// Adds an event listener to this element by id.
-	/// @lifetime The added listener must stay alive until after the dispatched call from EventListener::OnDetach(). This occurs
-	///     eg. when the element is destroyed or when RemoveEventListener() is called with the same parameters passed here.
+	/// @warning The added listener must stay alive until after the dispatched call from EventListener::OnDetach().
+	/// This occurs eg. when the element is destroyed or when RemoveEventListener() is called with the same parameters passed here.
```
Comment (mikke89) [inline Include/RmlUi/Core/Element.h:480]: So these `@lifetime` notes are used consistently throughout the library, I say we revert these changes completely.

**Review by mikke89 — COMMENTED** [review]
Yeah, this approach should work fine as far as I can tell.

Could you perhaps add some unit tests as well? Thanks!

**Comment by ZombieRaccoon** [comment]
> Yeah, this approach should work fine as far as I can tell.
> 
> Could you perhaps add some unit tests as well? Thanks!

Absolutely! Your original request didn't go unheard, it's just that I didn't manage to fit both into my time. I'll make sure to follow up with unit tests.

File: Include/RmlUi/Core/Element.h (hunk 4 above)
Comment (ZombieRaccoon) [reply Include/RmlUi/Core/Element.h:480]: The reason I went for the change is that ```lifetime``` is not a valid Doxygen / Doxypress command and the meaning it seems to convey is along the lines of a ```warning```. With that being said, I shall revert back to the original if that's what you wish.

Comment (mikke89) [reply Include/RmlUi/Core/Element.h]: I understand, I'm not really too worried about that though as we don't actually use Doxygen for documentation.

**Comment by ZombieRaccoon** [comment]
@mikke89 have you thought about using a mocking framework? I'm in the midst of writing the unit tests and it'd come in quite handy. Seeing as the project is using doctest as its unit testing framework, I'd suggest [trompeloeil](https://github.com/rollbear/trompeloeil) based on some research on the matter. If you were to be in agreement, I'd integrate the library under `Tests\Dependencies`.

**Comment by mikke89** [comment]
I don't really have any experience with mocking frameworks. I'm not really convinced by the reading I did on it, but if you think we can express some tests better I'll trust you on that. The library you mention looks good in terms of integration and licensing.

**Comment by ZombieRaccoon** [comment]
In my opinion, the most important thing that mocking frameworks save you is writing custom fake implementations like the one I resorted to in ```FakeEventListenerInstancer.h```. The main goal of such constructs is to verify certain dynamic characteristics of software - whether the ```InstanceEventListener``` method of the registered ```EventListenerInstancer``` implementation is invoked with certain parameters, in this particular case - that would be extremely tedious work otherwise because of the inherent opaqueness of object-oriented architectures from certain points of entry.

I'll implement the same thing using the library I've mentioned and you'll immediately see the difference in readability and likely recognize possible future use or even potential test refactors stemming from the opportunities its expressive power provides.

**Comment by mikke89** [comment]
I see, please go ahead with it, I'm eager to see the results. Good work on the tests so far, thanks for the thorough work!

**Comment by mikke89** [comment]
There you go, compiler finally gives in ;)

Let me know when you feel happy with it by marking it as ready. Could you clean up the commit history? I'm thinking one commit for the library integration, one or two for the core changes, and one or two for the unit tests. Thanks!

**Comment by ZombieRaccoon** [comment]
Yeah, that warning from GCC was interesting to say the least. Anyway, I believe that the deed is done. I do agree that the commit history got a little cluttered, so I cleaned it up.

**Review by mikke89 — APPROVED** [review]

**Comment by mikke89** [comment]
Looks good, thanks again!
